### ISSUE TITLE
JavaScript exception when starting video

### ISSUE THREAD

**Issue by AnDH2705** [0xced/XCDYouTubeKit#193]
[XCDYouTubeKit] JavaScript exception: TypeError: undefined is not an object
[XCDYouTubeKit] Video operation finished with error: The operation couldn’t be completed. (XCDYouTubeVideoErrorDomain error -2.)

It just happen a few minutes ago. I guess because of some update from Youtube's side.

**Comment by 0xced** [comment]
Can you please tell the exact identifier of the video you tried to play? I think this issue might be related to #194.

**Comment by codaman** [comment]
B3eAMGXFw1o this video also. age restriction problem

**Comment by AnDH2705** [comment]
Yes it is related to #194. The error happens when certain videos are played.

**Comment by patrickabadi** [comment]
I can see that the code in playerScript is no longer valid with the new js file from youTube.

This is the code:
static NSString *jsPrologue = @"(function()";
	static NSString *jsEpilogue = @")();";
	if ([script hasPrefix:jsPrologue] && [script hasSuffix:jsEpilogue])

Looking at the js file it no longer starts with "(function()".

This is what it looks like now:

var _yt_player={};(function(g){var window=this;var f;v.... 

so you would have to strip out everything before "var window..."

Also, the ending of the js file looks like this:

&$$("pe");})(_yt_player);

So basically ")(_yt_player);" needs to be stripped out.

**Comment by RVijay007** [comment]
@patrickabadi Has this worked for you. For me, this alone does not seem to be a sufficient solution. The script itself seems to have changed.

**Comment by patrickabadi** [comment]
Hello,

 

Yes it does.  I’m doing my project in C# Xamarin so my implementation is very different.  One of the big things that I’m doing differently that has been a huge help for me is I don’t use the straight javascript engine and try and fake the DOM, I download the javascript code and place it in a hidden UIWebview and then call the evaluatejavascript:

 

public void HandleJavascriptResponse(string response, Action<bool> fnResponse)

              {

                     try

                     {

                           javascript = PlayerScript.StripFunctionHeader(response);

                           fnJavascriptResponse = fnResponse;

 

                           string html = 

                                  "<!DOCTYPE html>" +

                                  "<html lang=\"en\" xmlns=\"http://www.w3.org/1999/xhtml\">" +

                                  "<head>" +

                                  "\t<meta charset=\"utf-8\" />" +

                                  "\t<title>Video Page</title>" +

                                  "\t<script lang=\"en-us\" type=\"text/javascript\" >" +

                                  "\t{0}" +

                                  "\t</script>" +

                                  "</head>" +

                                  "<body></body>" +

                                  "</html>";

 

                           html = string.Format(html, javascript);

 

                           ImageLoader.nsDispatcher.InvokeOnMainThread(()=>

                                  {

                                         Debug.WriteLine("Creating webview");

                                         web = new UIWebView();

 

                                         web.LoadFinished += (object sender, EventArgs e) => 

                                         {

                                                Debug.WriteLine("Finished loading webview");

 

                                                playerScript = new PlayerScript(javascript, web);

 

                                                HandleVideoInfoWithInfo(videoInfo, fnJavascriptResponse);

                                         };

 

                                         web.LoadHtmlString(html, new NSUrl("http://www.panoramicsoft.com"));

                                  });                        

                     }

                     catch(Exception exp)

                     {

                           Debug.WriteLine("HandleJavascriptResponse exception: " + exp);

                           playerScript = null;

                           fnResponse(false);

                           return;

                     }

 

 

 

              }

 

I’ve included my PlayerScript class so you can see how I use the UIWebview to unscramble the signature.

 

Let me know if this helps!

 

Patrick

**Comment by 0xced** [comment]
I just released version 2.3.3 which addresses this issue.

**Comment by nvrtdfrst** [comment]
Thanks for the quick turnaround @0xced !

**Comment by codaman** [comment]
very thanks @0xced

**Comment by tulurira** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

tulurira3@naver.com 님께 보내신 메일 <Re: [XCDYouTubeKit] JavaScript exception when starting video (#193)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by vrdriver** [comment]
Hi @0xced, 
I'm seeing this message as well when a video has started:
"[XCDYouTubeKit] JavaScript exception: TypeError: undefined is not an object (evaluating 'window.location.hostname.split')"

Any thoughts?
Cheers,
Steve
